### PR TITLE
Adding error details display

### PR DESCRIPTION
**Pull request by gaelL**
This pull request add the display of error details from Cycloid API

closes: https://github.com/cycloidio/cycloid-cli/issues/11

### CODE REVIEW

**Comment by gaelL** [comment]
Rebased on develop and running go imports on all files

File: cmd/cycloid/pipelines/get-job.go [hunk 1]
```diff
@@ -58,17 +58,21 @@ func getJob(cmd *cobra.Command, args []string) error {
...
 		return errors.Wrap(err, "unable to get printer")
 	}
 
+	j, err := m.GetPipelineJob(org, project, env, job)
+	if err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
+		return errors.Wrap(err, "unable to get job")
```
Comment (sdurrheimer) [inline cmd/cycloid/pipelines/get-job.go:73]:
```suggestion
		return errors.Wrapf(err, "unable to get job: %s", job)
```

File: cmd/cycloid/pipelines/diff.go [hunk 2]
```diff
@@ -69,16 +84,17 @@ func diff(cmd *cobra.Command, args []string) error {
 
 	pd, err := m.DiffPipeline(org, project, env, pipelineTemplate, vars)
 	if err != nil {
-		return err
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
+		return errors.Wrap(err, "unable obtain a diff")
```
Comment (sdurrheimer) [inline cmd/cycloid/pipelines/diff.go:91]:
```suggestion
		return errors.Wrapf(err, "unable obtain pipeline's diff: %s-%s", project, env)
```

File: cmd/cycloid/pipelines/clear-task-cache.go [hunk 3]
```diff
@@ -53,8 +57,22 @@ func cleartaskCache(cmd *cobra.Command, args []string) error {
...
+	if err != nil {
+		return errors.Wrap(err, "unable to get printer")
+	}
 
 	if err := m.ClearTaskCachePipeline(org, project, env, job, task); err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
 		return errors.Wrap(err, "unable to clear task cache")
```
Comment (sdurrheimer) [inline cmd/cycloid/pipelines/clear-task-cache.go:76]:
```suggestion
		return errors.Wrapf(err, "unable to clear task's cache: %s", task)
```

File: cmd/cycloid/organizations/get.go [hunk 4]
```diff
@@ -42,17 +42,21 @@ func get(cmd *cobra.Command, args []string) error {
...
 		return errors.Wrap(err, "unable to get printer")
 	}
 
+	o, err := m.GetOrganization(org)
+	if err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
+		return errors.Wrap(err, "unable to get organization")
```
Comment (sdurrheimer) [inline cmd/cycloid/organizations/get.go:57]:
```suggestion
		return errors.Wrapf(err, "unable to get organization: %s", org)
```

File: cmd/cycloid/organizations/delete.go [hunk 5]
```diff
@@ -35,7 +39,22 @@ func del(cmd *cobra.Command, args []string) error {
...
+	if err != nil {
+		return errors.Wrap(err, "unable to get printer")
+	}
+
 	if err := m.DeleteOrganization(org); err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
 		return errors.Wrap(err, "unable to delete organization")
```
Comment (sdurrheimer) [inline cmd/cycloid/organizations/delete.go:58]:
```suggestion
		return errors.Wrapf(err, "unable to delete organization: %s", org)
```

File: cmd/cycloid/organizations/create.go [hunk 6]
```diff
@@ -44,17 +44,21 @@ func create(cmd *cobra.Command, args []string) error {
...
 		return errors.Wrap(err, "unable to get printer")
 	}
 
+	o, err := m.CreateOrganization(name)
+	if err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
+		return errors.Wrap(err, "unable to create organization")
```
Comment (sdurrheimer) [inline cmd/cycloid/organizations/create.go:59]:
```suggestion
		return errors.Wrapf(err, "unable to create organization: %s", name)
```

File: cmd/cycloid/kpis/delete.go [hunk 7]
```diff
@@ -39,8 +43,22 @@ func delete(cmd *cobra.Command, args []string) error {
...
+	if err != nil {
+		return errors.Wrap(err, "unable to get printer")
+	}
 
 	if err = m.DeleteKpi(org, can); err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
 		return errors.Wrap(err, "unable to delete kpi")
```
Comment (sdurrheimer) [inline cmd/cycloid/kpis/delete.go:62]:
```suggestion
		return errors.Wrapf(err, "unable to delete KPI: %s", can)
```

File: cmd/cycloid/infrapolicies/validate.go [hunk 8]
```diff
@@ -59,19 +60,23 @@ func validate(org, project, env, planPath, output string) error {
...
+
 	api := common.NewAPI()
 	m := middleware.NewMiddleware(api)
 	res, err := m.ValidateInfraPolicies(org, project, env, plan)
 	if err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
 		return fmt.Errorf("unable to validate terraform plan file: %w", err)
```
Comment (sdurrheimer) [inline cmd/cycloid/infrapolicies/validate.go:77]:
```suggestion
		return errors.Wrap(err, "unable to validate terraform plan file")
```

File: cmd/cycloid/external-backends/delete.go [hunk 9]
```diff
@@ -39,7 +43,21 @@ func del(cmd *cobra.Command, args []string) error {
...
+	if err != nil {
+		return errors.Wrap(err, "unable to get printer")
+	}
+
 	if err := m.DeleteExternalBackend(org, id); err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
 		return errors.Wrap(err, "unable to delete external backend")
```
Comment (sdurrheimer) [inline cmd/cycloid/external-backends/delete.go:61]:
```suggestion
		return errors.Wrapf(err, "unable to delete external backend: %d", id)
```

File: cmd/cycloid/creds/get.go [hunk 10]
```diff
@@ -51,17 +51,21 @@ func get(cmd *cobra.Command, args []string) error {
...
 		return errors.Wrap(err, "unable to get printer")
 	}
 
+	c, err := m.GetCredential(org, can)
+	if err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
+		return errors.Wrap(err, "unable to get credential")
```
Comment (sdurrheimer) [inline cmd/cycloid/creds/get.go:66]:
```suggestion
		return errors.Wrapf(err, "unable to get credential: %s", can)
```

File: cmd/cycloid/creds/delete.go [hunk 11]
```diff
@@ -37,8 +41,22 @@ func del(cmd *cobra.Command, args []string) error {
...
+	if err != nil {
+		return errors.Wrap(err, "unable to get printer")
+	}
 
 	if err := m.DeleteCredential(org, can); err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
 		return errors.Wrap(err, "unable to delete credential")
```
Comment (sdurrheimer) [inline cmd/cycloid/creds/delete.go:60]:
```suggestion
		return errors.Wrapf(err, "unable to delete credential: %s", can)
```

File: cmd/cycloid/creds/create.go [hunk 12]
```diff
@@ -354,6 +368,10 @@ func create(cmd *cobra.Command, args []string) error {
 	}
 
 	if err := m.CreateCredential(org, name, credT, rawCred, path, description); err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
 		return errors.Wrap(err, "unable to create credential")
```
Comment (sdurrheimer) [inline cmd/cycloid/creds/create.go:375]:
```suggestion
		return errors.Wrapf(err, "unable to create credential: %s", name)
```

File: cmd/cycloid/config-repositories/update.go [hunk 13]
```diff
@@ -75,22 +75,27 @@ func updateConfigRepository(cmd *cobra.Command, args []string) error {
...
 		return errors.Wrap(err, "unable to get printer")
 	}
 
+	cr, err := m.UpdateConfigRepository(org, can, cred, name, url, branch, setDefault)
+	if err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
+		return errors.Wrap(err, "unable to update config repository")
```
Comment (sdurrheimer) [inline cmd/cycloid/config-repositories/update.go:96]:
```suggestion
		return errors.Wrapf(err, "unable to update config repository: %s", can)
```

File: cmd/cycloid/config-repositories/delete.go [hunk 14]
```diff
@@ -44,7 +44,22 @@ func deleteConfigRepository(cmd *cobra.Command, args []string) error {
...
+	if err != nil {
+		return errors.Wrap(err, "unable to get printer")
+	}
+
 	if err := m.DeleteConfigRepository(org, can); err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
 		return errors.Wrap(err, "unable to delete config repository")
```
Comment (sdurrheimer) [inline cmd/cycloid/config-repositories/delete.go:63]:
```suggestion
		return errors.Wrapf(err, "unable to delete config repository: %s", can)
```

File: cmd/cycloid/catalog-repositories/update.go [hunk 15]
```diff
@@ -77,17 +73,21 @@ func updateCatalogRepository(cmd *cobra.Command, args []string) error {
...
 		return errors.Wrap(err, "unable to get printer")
 	}
 
+	cr, err := m.UpdateCatalogRepository(org, can, name, url, branch, cred)
+	if err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
+		return errors.Wrap(err, "unable to update catalog repository")
```
Comment (sdurrheimer) [inline cmd/cycloid/catalog-repositories/update.go:88]:
```suggestion
		return errors.Wrapf(err, "unable to update catalog repository: %s", can)
```

File: cmd/cycloid/catalog-repositories/delete.go [hunk 16]
```diff
@@ -42,7 +42,22 @@ func deleteCatalogRepository(cmd *cobra.Command, args []string) error {
...
+	if err != nil {
+		return errors.Wrap(err, "unable to get printer")
+	}
+
 	if err := m.DeleteCatalogRepository(org, canonical); err != nil {
+		// print the result on the standard output
+		if err := p.Print(err, printer.Options{}, os.Stdout); err != nil {
+			return errors.Wrap(err, "unable to print result")
+		}
 		return errors.Wrap(err, "unable to delete repository")
```
Comment (sdurrheimer) [inline cmd/cycloid/catalog-repositories/delete.go:61]:
```suggestion
		return errors.Wrapf(err, "unable to delete repository: %s", canonical)
```

**Review by sdurrheimer — COMMENTED** [review]
Some suggestions

File: cmd/cycloid/kpis/list-available.go [hunk 17]
```diff
@@ -42,17 +42,21 @@ func listAvailable(cmd *cobra.Command, args []string) error {
...
-		return err
-	}
-
 	// fetch the printer from the factory
 	p, err := factory.GetPrinter(output)
 	if err != nil {
 		return errors.Wrap(err, "unable to get printer")
 	}
 
+	kpis, err := m.ListAvailableKpi(org)
```
Comment (sdurrheimer) [inline cmd/cycloid/kpis/list-available.go:51]: [getAvailableKPIs](https://github.com/cycloidio/youdeploy-http-api/pull/2603) is going to be removed, maybe we can remove it right away.

File: cmd/cycloid/projects/create.go [hunk 18]
```diff
@@ -116,10 +122,12 @@ func create(cmd *cobra.Command, args []string) error {
 	vars := string(rawVars)
 
 	project, err := m.CreateProject(org, name, canonical, env, pipelineTemplate, vars, description, stackRef, usecase, configRepo)
+	err = printer.SmartPrint(p, nil, err, "unable to create pipeline", printer.Options{}, os.Stdout)
```
Comment (sdurrheimer) [inline cmd/cycloid/projects/create.go:125]:
```suggestion
	err = printer.SmartPrint(p, nil, err, "unable to create project", printer.Options{}, os.Stdout)
```

File: printer/printer.go [hunk 19]
```diff
@@ -1,9 +1,32 @@
 package printer
 
-import "io"
-
+import (
+	"io"
+	"github.com/pkg/errors"
```
Comment (sdurrheimer) [inline printer/printer.go:5]:
```suggestion
	"io"

	"github.com/pkg/errors"
```

File: printer/table/printer.go [hunk 20]
```diff
@@ -10,6 +10,7 @@ import (
 
 	"github.com/cycloidio/cycloid-cli/printer"
 	"github.com/olekukonko/tablewriter"
+	"github.com/cycloidio/cycloid-cli/client/models"
```
Comment (sdurrheimer) [inline printer/table/printer.go:13]:
```suggestion
	"github.com/olekukonko/tablewriter"
	
	"github.com/cycloidio/cycloid-cli/client/models"
	"github.com/cycloidio/cycloid-cli/printer"
```

File: main.go [hunk 21]
```diff
@@ -21,15 +20,23 @@ var (
...
 )
 
+func inRed(msg string) string {
+	return fmt.Sprintf("\033[1;31m%s\033[0m", msg)
+}
+
 func Execute() {
 	if err := rootCmd.Execute(); err != nil {
+		rootCmd.PrintErrln(inRed("Error:"), err.Error())
+		// rootCmd.PrintErrf("Run '%v --help' for usage.\n", rootCmd.CommandPath())
```
Comment (xescugc) [inline main.go:39]: comment

**Review by xescugc — COMMENTED** [review]
IDK if it was commented, just picked a bit but i saw removing the `%w` which is the new format for GO (standard lib) handles wrapping errors.

Also I would encourage using `xerrors` as it's more of the actual direction that GO is going in terms of formatting errors, also mentioned on https://github.com/cycloidio/cycloid-cli/issues/11#issuecomment-694848182.

For what I see you only wrap things up and down so I don't see the issue on having `%w` being use instead of having a lib do it and not do it the way it's suppose to be done in GO.

But then again I did not follow this project so maybe there is something else IDK hehe.

File: main.go (hunk 21 above)
Comment (gaelL) [reply main.go:39]: Note discussed internally and we will merge this PR to provide a display of error details as it is blocking for our users.
We will then have a second iteration to redefine the error wrapping.

The goal would be to be able to detect if it's a Cobra (command syntax) error or an error from Cycloid API.
To display only the  `--help` when it's an issue from the command line

**Pull request merged**